ui/src/app/pages/workspace/workspace-library.tsx

```diff
@@ -67,55 +64,78 @@ const LibraryTab: React.FunctionComponent<{
     </Clickable>;
   };
 
+class ProfileState {
+  profile: Profile;
+  reload: Function;
+}
+
+class CurrentTab {
+  title: string;
+  icon: string;
+}
+
+class Props {
+  profileState: ProfileState;
+  enablePublishedWorkspaces: boolean;
+}
+
+class State {
+  currentTab: CurrentTab;
+  errorText: string;
+  workspaceList: WorkspacePermissions[];
+  publishedWorkspaces: WorkspacePermissions[];
+  featuredWorkspaces: WorkspacePermissions[];
+  workspacesLoading: boolean;
+}
+
 export const WorkspaceLibrary = withUserProfile()
-(class extends React.Component<
-  {profileState: { profile: Profile, reload: Function } },
-  {currentTab: {title: string, icon: string}, errorText: string,
-    workspaceList: WorkspacePermissions[], featuredWorkspaces: WorkspacePermissions[],
-    workspacesLoading: boolean}> {
+(class extends React.Component<Props, State> {
   constructor(props) {
     super(props);
     this.state = {
       currentTab: libraryTabEnums.FEATURED_WORKSPACES,
       errorText: '',
       featuredWorkspaces: [],
+      publishedWorkspaces: [],
       workspaceList: [],
       workspacesLoading: true
     };
   }
 
+  libraryTabs = (this.props.enablePublishedWorkspaces || environment.enablePublishedWorkspaces)
```

Comment:
This boolean logic feels a bit awkward – why do we need to look at both the props *and* the environment to decide whether to enable showing published workspaces? I'd be inclined to choose one or the other. If (as I might guess) you're using props to enable easier testing, I'd recommend aligning entirely on that pattern.

Reply:
I meant, if props is needed for testing, just use the props value and don't refer to the environment at all from here.